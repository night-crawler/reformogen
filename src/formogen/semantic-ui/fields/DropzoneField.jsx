import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { Button, Checkbox, Form, Icon, Image, List, Segment } from 'semantic-ui-react';
import Measure from 'react-measure';
import Dropzone from 'react-dropzone';

import { errorsType } from '../../fieldPropTypes';
import { fileTypeImageMapping, UNKNOWN_FILE_TYPE } from '../../fileTypeImageMapping';
import { splitExt, bytesToSize } from '../../utils';
import ErrorsList from '../common/ErrorsList';
import CaptionTruncator from '../common/CaptionTruncator';

const fileShape = PropTypes.shape({
  lastModified: PropTypes.number,
  lastModifiedDate: PropTypes.object,
  preview: PropTypes.string,
  size: PropTypes.number,
  type: PropTypes.string,
});


ListContentWithProgress.propTypes = {
  percent: PropTypes.number,
  color: PropTypes.string,
  styles: PropTypes.object,
  children: PropTypes.any,
};
function ListContentWithProgress({ percent = 50, color = 'LimeGreen', styles = {}, children = {} }) {
  const defaultStyles = {
    background: `linear-gradient(90deg, #fff, ${color} ${percent}%, #fff ${percent}%, #fff 100%)`,
    backgroundRepeat: 'no-repeat',
    border: '1px solid #999',
    borderRadius: 4,  // don't touch this it's a magic number!
    height: '37px',
  };
  return (
    <List.Content className='list-content-with-progress' style={ { ...defaultStyles, styles } }>
      { children }
    </List.Content>
  );
}


/**
 * {
 *     formFilesUploadProgress: {
 *         <fieldName>: {
 *              <fileName>: {percent: 100, status: 'ok|fail'}
 *         }
 *     }
 * }
 */

FilesPreview.propTypes = {
  files: PropTypes.arrayOf(fileShape).isRequired,
  onClear: PropTypes.func.isRequired,
  onRemoveFile: PropTypes.func.isRequired,
  getFileIcon: PropTypes.func.isRequired,
  formFilesUploadProgress: PropTypes.object
};
function FilesPreview({ files, onClear, getFileIcon, onRemoveFile, formFilesUploadProgress }) {
  if (_.isEmpty(files))
    return null;

  return (
    <Segment attached={ true }>
      <List divided={ true } relaxed={ true } className='files-previews'>
        { files.map((file, i) => {
          const progress = _.get(formFilesUploadProgress, file.name, {});
          return <FileItem
            key={ i }
            file={ file }
            getFileIcon={ getFileIcon }
            onRemove={ onRemoveFile }
            progress={ progress }
          />;
        }) }
      </List>
      <Button size='mini' fluid={ true } icon={ true } onClick={ onClear }>
        <Icon name='remove' />
      </Button>
    </Segment>
  );
}


CurrentFilesPreview.propTypes = {
  currentFiles: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),
  removable: PropTypes.bool,
  onRemove: PropTypes.func,
};
function CurrentFilesPreview({ currentFiles, removable = true, onRemove }) {
  const files = _.isArray(currentFiles) ? currentFiles : [currentFiles];

  const listItems = files.map((value, i) => (
    <List.Item key={ i }>
      <a href={ value } target='_blank' rel='noopener noreferrer'>
        { typeof value === 'string' ? value.split('/').slice(-1) : 'File' }
      </a>
    </List.Item>
  ));

  return (
    <Segment attached={ true }>
      <List>
        <List.Header>Currently:</List.Header>
        <List.Item>
          <List.Content floated='right'>
            { removable && <Checkbox label='clear' floated='right' onChange={ onRemove } /> }
          </List.Content>
          <List bulleted={ true }>{ listItems }</List>
        </List.Item>
      </List>
    </Segment>
  );
}


FileItem.propTypes = {
  file: fileShape.isRequired,
  getFileIcon: PropTypes.func.isRequired,
  onRemove: PropTypes.func.isRequired,
  progress: PropTypes.object.isRequired,
};
function FileItem({ file, getFileIcon, onRemove, progress }) {
  const [fileName, fileExt] = splitExt(file.name);
  const { status: uploadStatus, percent = 0 } = progress;
  const color = uploadStatus === 'fail' ? '#db2828' : 'LimeGreen';

  return (
    <List.Item>
      <Measure bounds={ true }>
        {
          ({ measureRef, contentRect }) =>
            <div ref={ measureRef }>
              <ListContentWithProgress percent={ percent } color={ color }>
                <Image verticalAlign='middle' size='mini' src={ getFileIcon(file) } floated='left' />
                <div className='file-representation' title={ `[${bytesToSize(file.size)}] ${file.name}` }>
                  <strong>{ `[${bytesToSize(file.size)}]` }&nbsp;</strong>
                  <CaptionTruncator caption={ fileName } width={ (contentRect.bounds.width || 200) - 200 } />
                  <span>.{ fileExt }</span>
                </div>
                <Button icon={ true } size='mini' attached='right' floated='right'
                  onClick={ () => onRemove(file) }>
                  <Icon name='remove' />
                </Button>
              </ListContentWithProgress>
            </div>
        }
      </Measure>
    </List.Item>
  );
}


export default class DropzoneField extends React.Component {
    static propTypes = {
      type: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      verbose_name: PropTypes.string.isRequired,
      help_text: PropTypes.string.isRequired,
      upperFirstLabel: PropTypes.bool,
      required: PropTypes.bool,
      editable: PropTypes.bool,
      updateProps: PropTypes.func,
      widget: PropTypes.string,
      errors: errorsType,

      upload_url: PropTypes.string,
      delete_url: PropTypes.string,

      onChange: PropTypes.func.isRequired,
      getFileIcon: PropTypes.func,
      multiple: PropTypes.bool,
      accept: PropTypes.string,

      layoutOpts: PropTypes.object,
      formFilesUploadProgress: PropTypes.object,

      value: PropTypes.string,
    };
    static defaultProps = {
      getFileIcon: (fileObject) => {
        let ext = fileObject.name.split('.').pop().toLowerCase();
        return fileTypeImageMapping[ext] || UNKNOWN_FILE_TYPE;
      },
      multiple: false,
    };

    constructor(props) {
      super(props);
      this.state = { files: [] };
    }

    handleRemoveCurrentFiles = (event, { checked }) => {
      if (checked) {
        return this.props.onChange(
          null,
          {
            name: this.props.name,
            value: { files: [{ name: this.props.value }], url: this.props.delete_url, action: 'delete' }
          }
        );
      }

      return this.props.onChange(
        null,
        {
          name: this.props.name,
          value: { files: [], url: this.props.upload_url, action: 'upload' }
        }
      );
    };
    handleDrop = (newFiles) => {
      const { files: oldFiles } = this.state;

      if (!this.props.multiple) {
        this.setState({ files: newFiles });
        return this.props.onChange(
          null,
          {
            name: this.props.name,
            value: { files: newFiles, url: this.props.upload_url, action: 'upload' }
          }
        );
      }

      // build key by this triple to ensure we're adding different newFiles
      const keyBuilder = (item) => `${ item.lastModified }:${ item.size }:${ item.name }`;
      const mergedFiles = _(oldFiles)
        .keyBy(keyBuilder)
        .merge(_.keyBy(newFiles, keyBuilder))
        .values()
        .value();

      this.setState({ files: mergedFiles });
      return this.props.onChange(
        null,
        {
          name: this.props.name,
          value: { files: mergedFiles, url: this.props.upload_url, action: 'upload' }
        }
      );
    };
    handleClearFiles = () => {
      this.setState({ files: [] });
      this.props.onChange(
        null,
        {
          name: this.props.name,
          value: { files: [], url: this.props.upload_url, action: 'upload' }
        }
      );
    };
    handleRemoveFile = (fileObject) => {
      const files = _.without(this.state.files, fileObject);
      this.setState({ files });

      return this.props.onChange(
        null,
        {
          name: this.props.name,
          value: { files, url: this.props.upload_url, action: 'upload' }
        }
      );
    };

    render() {
      const labelText = this.props.upperFirstLabel ? _.upperFirst(this.props.verbose_name) : this.props.verbose_name;

      let _props = {
        multiple: this.props.multiple,
        onDrop: this.handleDrop,
        accept: this.props.accept,
      };

      if (_.isFunction(this.props.updateProps)) {
        _props = this.props.updateProps(_props, this.props);
      }

      const showCurrentFilesPreview = this.props.value && _.isEmpty(this.state.files);

      return (
        <Form.Field
          required={ this.props.required }
          disabled={ !this.props.editable }
          width={ this.props.layoutOpts.width }
          error={ !_.isEmpty(this.props.errors) }
        >
          <Dropzone className='ui center aligned dropzone segment attached top'  { ..._props }>
            <strong>{ labelText } { this.props.required && <span className='ui red'>*</span> }</strong>
            { this.props.help_text && <div className='help-text'>{ this.props.help_text }</div> }
            <div>
              { !_.isEmpty(this.state.files) || 'Drop files here' }
            </div>
          </Dropzone>

          <FilesPreview
            files={ this.state.files }
            onClear={ this.handleClearFiles }
            getFileIcon={ this.props.getFileIcon }
            onRemoveFile={ this.handleRemoveFile }
            formFilesUploadProgress={ this.props.formFilesUploadProgress }
          />

          {
            showCurrentFilesPreview &&
            <CurrentFilesPreview
              currentFiles={ this.props.value }
              removable={ !!this.props.delete_url }
              onRemove={ this.handleRemoveCurrentFiles }
            />
          }

          <ErrorsList messages={ this.props.errors } />
        </Form.Field>
      );
    }
}
