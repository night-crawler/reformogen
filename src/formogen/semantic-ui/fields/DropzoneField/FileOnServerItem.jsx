import React from 'react';
import PropTypes from 'prop-types';
import { List, Image, Button } from 'semantic-ui-react';

import { splitExt } from '../../../utils';
import { fileTypeImageMapping, UNKNOWN_FILE_TYPE } from '../../../fileTypeImageMapping';



FileOnServerItem.propTypes = {
  fileUrl: PropTypes.string,
  getPreviewImage: PropTypes.func,
  onDelete: PropTypes.func,
  isRemovable: PropTypes.bool,
};
FileOnServerItem.defaultProps = {
  getPreviewImage: fileUrl => 
    fileTypeImageMapping[ splitExt(fileUrl)[1] ] || UNKNOWN_FILE_TYPE,
  onDelete: () => // eslint-disable-next-line
    console.warn('FileOnServerItem.onDelete'),
};
export function FileOnServerItem(props) {
  const fileName = props.fileUrl.split('/').pop();

  // TODO: watermelon come here, motherfucker!
  return (
    <List.Item>
      <Image 
        src={ props.getPreviewImage(props.fileUrl) } 
        verticalAlign='middle' size='mini'
      />
      <a href={ props.fileUrl }>{ fileName }</a>
      
      { props.isRemovable && <Button 
        icon='remove' size='mini' attached='right' floated='right' 
        onClick={ props.onDelete }
      /> }
    </List.Item>
  );
}
