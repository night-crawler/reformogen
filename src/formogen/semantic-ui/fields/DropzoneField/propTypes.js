import PropTypes from 'prop-types';


export const fileShape = {
  lastModified: PropTypes.number,
  lastModifiedDate: PropTypes.object,
  name: PropTypes.string.isRequired,
  size: PropTypes.number.isRequired,
  type: PropTypes.string,
};

export const fileType = PropTypes.shape(fileShape);
