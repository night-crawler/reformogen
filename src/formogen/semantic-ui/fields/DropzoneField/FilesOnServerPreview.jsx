import React from 'react';
import PropTypes from 'prop-types';
import { isArray } from 'lodash';
import { List, Segment } from 'semantic-ui-react';

import { FileOnServerItem } from './FileOnServerItem';


FilesOnServerPreview.propTypes = {
  isRemovable: PropTypes.bool,
  files: PropTypes.any,
  onDeleteFile: PropTypes.func,
};
FilesOnServerPreview.defaultProps = {
  isRemovable: false,
  files: [],
  onDeleteFile: (fileUrl) => // eslint-disable-next-line
    console.warn('FilesOnServerPreview.onDeleteFile', fileUrl),
};
export function FilesOnServerPreview(props) {
  // TODO: it looks like a pipe dream, there's no way for multiple files to be stored in one field
  const files = isArray(props.files) 
    ? props.files 
    : [ props.files ];

  return (
    <Segment as={ List } attached='bottom' divided={ true } relaxed={ true }>
      { files.map((fileUrl, i) => 
        <FileOnServerItem 
          key={ i } 
          fileUrl={ fileUrl } 
          isRemovable={ props.isRemovable }
          onDelete={ props.onDeleteFile }  
        />
      ) }
    </Segment>
  );
}
