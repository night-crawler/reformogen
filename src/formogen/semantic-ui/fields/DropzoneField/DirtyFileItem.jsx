import React from 'react';
import PropTypes from 'prop-types';
import { List, Image, Button } from 'semantic-ui-react';

import { splitExt, bytesToSize } from '../../../utils';
import { fileTypeImageMapping, UNKNOWN_FILE_TYPE } from '../../../fileTypeImageMapping';
import { CaptionTruncator } from '../../CaptionTruncator';

import { fileType } from './propTypes';


DirtyFileItem.propTypes = {
  file: fileType,
  getPreviewImage: PropTypes.func,
  onDelete: PropTypes.func,
};
DirtyFileItem.defaultProps = {
  getPreviewImage: file => 
    fileTypeImageMapping[ splitExt(file.name)[1] ] || UNKNOWN_FILE_TYPE,
  onDelete: () => // eslint-disable-next-line
    console.warn('DirtyFileItem.onDelete'),
};
export function DirtyFileItem(props) {
  const [ fileName, fileExt ] = splitExt(props.file.name);
  const sizeRepr = `[${bytesToSize(props.file.size)}]`;

  // TODO: watermelon come here, motherfucker!
  return (
    <List.Item>
      <Image 
        src={ props.getPreviewImage(props.file) } 
        verticalAlign='middle' size='mini'
      />

      <List.Content style={ { width: '100%' } }>
        <strong>{ sizeRepr }&nbsp;</strong>
        <CaptionTruncator caption={ fileName } width='50%' />
        <span>.{ fileExt }</span>
      </List.Content>

      <Button 
        icon='remove' size='mini' attached='right' floated='right' 
        onClick={ props.onDelete }
      />
    </List.Item>
  );
}
