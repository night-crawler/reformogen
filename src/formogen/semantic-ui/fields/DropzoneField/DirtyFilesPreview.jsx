import React from 'react';
import PropTypes from 'prop-types';
import { List, Segment, Button } from 'semantic-ui-react';

import { fileType } from './propTypes';
import { DirtyFileItem } from './DirtyFileItem';

DirtyFilesPreview.propTypes = {
  files: PropTypes.arrayOf(fileType),
  onClear: PropTypes.func,
  onDeleteFile: PropTypes.func,
};
DirtyFilesPreview.defaultProps = {
  files: [],
  onClear: () => // eslint-disable-next-line
    console.warn('DirtyFilesPreview.onClear'),
  onDeleteFile: (file, i) => // eslint-disable-next-line
    console.warn('DirtyFilesPreview.onDeleteFile', file, i),
};
export function DirtyFilesPreview(props) {
  return (
    <Segment as={ List } attached='bottom' divided={ true } relaxed={ true }>
      { props.files.map((file, i) => 
        <DirtyFileItem 
          key={ i } 
          file={ file } 
          onDelete={ () => props.onDeleteFile(file, i) } 
        />
      ) }
      <Button 
        size='mini' fluid={ true } icon='remove' 
        onClick={ props.onClear }
      />
    </Segment>
  );
}
