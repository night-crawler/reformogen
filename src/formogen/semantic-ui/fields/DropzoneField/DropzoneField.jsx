import React from 'react';
import PropTypes from 'prop-types';
import { Form } from 'semantic-ui-react';
import { isEmpty, upperFirst, keyBy, sortBy, reverse } from 'lodash';
import Dropzone from 'react-dropzone';

import { errorsType } from '../../../fieldPropTypes';
import { fileTypeImageMapping, UNKNOWN_FILE_TYPE } from '../../../fileTypeImageMapping';
import { ErrorsList } from '../../ErrorsList';

import { DirtyFilesPreview } from './DirtyFilesPreview';
import { FilesOnServerPreview } from './FilesOnServerPreview';


export class DropzoneField extends React.Component {
    static propTypes = {
      type: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      verbose_name: PropTypes.string.isRequired,
      help_text: PropTypes.string,
      upperFirstLabel: PropTypes.bool,
      required: PropTypes.bool,
      editable: PropTypes.bool,
      updateProps: PropTypes.func,
      widget: PropTypes.string,
      errors: errorsType,

      upload_url: PropTypes.string,
      delete_url: PropTypes.string,

      onChange: PropTypes.func.isRequired,
      getFileIcon: PropTypes.func,
      multiple: PropTypes.bool,
      accept: PropTypes.string,

      layoutOpts: PropTypes.object,
      formFilesUploadProgress: PropTypes.object,
      
      /** value is a current dirty value, previous is not here */
      value: PropTypes.array,
      initialValue: PropTypes.string,

      dropText: PropTypes.string,
    };
    static defaultProps = {
      initialValue: '',
      dropText: 'Drop files here',
      getFileIcon: (fileObject) => {
        let ext = fileObject.name.split('.').pop().toLowerCase();
        return fileTypeImageMapping[ext] || UNKNOWN_FILE_TYPE;
      },
      multiple: false,
      value: [],
    };

    constructor(props) {
      super(props);
      this.state = { files: [] };
    }

    handleRemoveCurrentFiles = (event, { checked }) => {
      if (checked) {
        return this.props.onChange(
          null,
          {
            name: this.props.name,
            value: { files: [{ name: this.props.value }], url: this.props.delete_url, action: 'delete' }
          }
        );
      }

      return this.props.onChange(
        null,
        {
          name: this.props.name,
          value: { files: [], url: this.props.upload_url, action: 'upload' }
        }
      );
    };
    handleDrop = newFiles => {
      if (!this.props.multiple)
        return this.props.onChange(null, {
          name: this.props.name,
          value: newFiles
        });

      // build key by this triple to ensure we're adding different newFiles
      const keyBuilder = item => `${ item.lastModified }:${ item.size }:${ item.name }`;
      const newUniqueFiles = {
        ...keyBy(this.props.value, keyBuilder),
        ...keyBy(newFiles, keyBuilder) 
      } |> Object.values 
        |> (values => sortBy(values, ['size']))
        |> reverse;
      
      return this.props.onChange(null, {
        name: this.props.name,
        value: newUniqueFiles
      });
    };
    handleClearFiles = () =>
      this.props.onChange(null, {
        name: this.props.name,
        value: []
      });

    handleRemoveFile = (fileObject, i) =>
      this.props.onChange(null, {
        name: this.props.name,
        value: this.props.value.filter((file, _i) => _i !== i)
      });

    handleRemoveFileFromServer = fileUrl => {
      // eslint-disable-next-line
      console.log(fileUrl);
    }

    render() {
      const labelText = this.props.upperFirstLabel 
        ? upperFirst(this.props.verbose_name) 
        : this.props.verbose_name;

      return (
        <Form.Field
          required={ this.props.required }
          disabled={ !this.props.editable }
          width={ this.props.layoutOpts.width }
          error={ !isEmpty(this.props.errors) }
        >
          <Dropzone 
            className='ui center aligned dropzone segment attached top'  
            multiple={ this.props.multiple }
            onDrop={ this.handleDrop }
            accept={ this.props.accept }
            acceptClassName='green'
            rejectClassName='red'
          >
            <strong>{ labelText } { this.props.required && <span className='ui red'>*</span> }</strong>
            { this.props.help_text && <div className='help-text'>{ this.props.help_text }</div> }
            <div>
              { !isEmpty(this.state.files) || this.props.dropText }
            </div>
          </Dropzone>

          <DirtyFilesPreview
            files={ this.props.value }
            onClear={ this.handleClearFiles }
            onDeleteFile={ this.handleRemoveFile }
          />

          <FilesOnServerPreview
            files={ this.props.initialValue }
            onDeleteFile={ this.handleRemoveFileFromServer }
            isRemovable={ !!this.props.delete_url }
          />
          <ErrorsList messages={ this.props.errors } />
        </Form.Field>
      );
    }
}
