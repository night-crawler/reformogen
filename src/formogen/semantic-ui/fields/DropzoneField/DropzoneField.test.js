import { DropzoneField } from './DropzoneField';


it('<DropzoneField />', () => {
  const onChangeFn = jest.fn();
  const wrapper = shallow(
    <DropzoneField
      type='DropzoneField'
      name='DropzoneField: name'
      verbose_name='DropzoneField: verbose_name'
      help_text='DropzoneField: help_text'
      displayOptions={ { width: 1 } }
      onChange={ onChangeFn }
    />
  );
  expect(wrapper).toMatchSnapshot();
});
