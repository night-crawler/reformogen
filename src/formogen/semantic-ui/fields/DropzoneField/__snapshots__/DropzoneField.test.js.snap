// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropzoneField /> 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DropzoneField
    displayOptions={
      Object {
        "width": 1,
      }
    }
    dropText="Drop files here"
    getFileIcon={[Function]}
    help_text="DropzoneField: help_text"
    initialValue=""
    multiple={false}
    name="DropzoneField: name"
    onChange={[MockFunction]}
    type="DropzoneField"
    value={Array []}
    verbose_name="DropzoneField: verbose_name"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Segment
          acceptClassName="green"
          as={[Function]}
          attached="top"
          className="dropzone"
          multiple={false}
          onDrop={[Function]}
          rejectClassName="red"
          textAlign="center"
        >
          <strong>
            DropzoneField: verbose_name
             
          </strong>
          <div
            className="help-text"
          >
            DropzoneField: help_text
          </div>
          <div>
            Drop files here
          </div>
        </Segment>,
        <DirtyFilesPreview
          files={Array []}
          onClear={[Function]}
          onDeleteFile={[Function]}
        />,
        <FilesOnServerPreview
          files=""
          isRemovable={false}
          onDeleteFile={[Function]}
        />,
        <ErrorsList
          color="red"
          header="Errors"
        />,
      ],
      "disabled": true,
      "error": false,
      "required": undefined,
      "width": 1,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "accept": undefined,
          "acceptClassName": "green",
          "as": [Function],
          "attached": "top",
          "children": Array [
            <strong>
              DropzoneField: verbose_name
               
            </strong>,
            <div
              className="help-text"
            >
              DropzoneField: help_text
            </div>,
            <div>
              Drop files here
            </div>,
          ],
          "className": "dropzone",
          "multiple": false,
          "onDrop": [Function],
          "rejectClassName": "red",
          "textAlign": "center",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "DropzoneField: verbose_name",
                " ",
                undefined,
              ],
            },
            "ref": null,
            "rendered": Array [
              "DropzoneField: verbose_name",
              " ",
              undefined,
            ],
            "type": "strong",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "DropzoneField: help_text",
              "className": "help-text",
            },
            "ref": null,
            "rendered": "DropzoneField: help_text",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Drop files here",
            },
            "ref": null,
            "rendered": "Drop files here",
            "type": "div",
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "files": Array [],
          "onClear": [Function],
          "onDeleteFile": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "files": "",
          "isRemovable": false,
          "onDeleteFile": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "color": "red",
          "header": "Errors",
          "messages": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Segment
            acceptClassName="green"
            as={[Function]}
            attached="top"
            className="dropzone"
            multiple={false}
            onDrop={[Function]}
            rejectClassName="red"
            textAlign="center"
          >
            <strong>
              DropzoneField: verbose_name
               
            </strong>
            <div
              className="help-text"
            >
              DropzoneField: help_text
            </div>
            <div>
              Drop files here
            </div>
          </Segment>,
          <DirtyFilesPreview
            files={Array []}
            onClear={[Function]}
            onDeleteFile={[Function]}
          />,
          <FilesOnServerPreview
            files=""
            isRemovable={false}
            onDeleteFile={[Function]}
          />,
          <ErrorsList
            color="red"
            header="Errors"
          />,
        ],
        "disabled": true,
        "error": false,
        "required": undefined,
        "width": 1,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "accept": undefined,
            "acceptClassName": "green",
            "as": [Function],
            "attached": "top",
            "children": Array [
              <strong>
                DropzoneField: verbose_name
                 
              </strong>,
              <div
                className="help-text"
              >
                DropzoneField: help_text
              </div>,
              <div>
                Drop files here
              </div>,
            ],
            "className": "dropzone",
            "multiple": false,
            "onDrop": [Function],
            "rejectClassName": "red",
            "textAlign": "center",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "DropzoneField: verbose_name",
                  " ",
                  undefined,
                ],
              },
              "ref": null,
              "rendered": Array [
                "DropzoneField: verbose_name",
                " ",
                undefined,
              ],
              "type": "strong",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "DropzoneField: help_text",
                "className": "help-text",
              },
              "ref": null,
              "rendered": "DropzoneField: help_text",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Drop files here",
              },
              "ref": null,
              "rendered": "Drop files here",
              "type": "div",
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "files": Array [],
            "onClear": [Function],
            "onDeleteFile": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "files": "",
            "isRemovable": false,
            "onDeleteFile": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "color": "red",
            "header": "Errors",
            "messages": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
